layer/compile_time/compile_time_layer.cc

```diff
@@ -59,39 +63,60 @@ class ShaderModuleSlackEvent : public Event {
                          Duration duration)
       : Event(name),
         shader_hash_("shader_hash", hash_value),
-        duration_("slack", duration) {
-    InitAttributes({&shader_hash_, &duration_});
+        duration_("slack", duration),
+        trace_attr_("trace_attr", kTraceEventCategory, "X",
+                    {&duration_, &shader_hash_}) {
+    InitAttributes({&shader_hash_, &duration_, &trace_attr_});
   }
 
  private:
   HashAttr shader_hash_;
   DurationAttr duration_;
+  TraceEventAttr trace_attr_;
 };
 
 class CreateShaderEvent : public Event {
  public:
   CreateShaderEvent(const char* name, int64_t hash_value, Duration duration)
       : Event(name),
         shader_hash_("shader_hash", hash_value),
-        duration_("duration", duration) {
-    InitAttributes({&shader_hash_, &duration_});
+        duration_("duration", duration),
+        trace_attr_("trace_attr", kTraceEventCategory, "X",
+                    {&duration_, &shader_hash_}) {
+    InitAttributes({&shader_hash_, &duration_, &trace_attr_});
   }
 
  private:
   HashAttr shader_hash_;
   DurationAttr duration_;
+  TraceEventAttr trace_attr_;
+};
+
+class CompileTimeStartEvent : public Event {
```

Comment:
I expected this event to denote compilation start but it seems like it's for layer initialization. Should we call it LayerInitEvent?

Reply:
in the future, we could move it to a headers so that it can be shared by all layers if we find it useful

Reply:
> I expected this event to denote compilation start but it seems like it's for layer initialization. Should we call it LayerInitEvent?

Sure!

Reply:
> in the future, we could move it to a headers so that it can be shared by all layers if we find it useful

That's a good idea. I've implemented this as a `DefaultInstantEvent` and it'll be in the next PR.